feat: 修复查询嵌套查询 --story=142619506 --bug=142619506

Zhaoyikaiii opened the pull request.

Zhaoyikaiii committed:
feat: 添加反向嵌套聚合支持，优化聚合信息处理逻辑

Zhaoyikaiii committed:
fix: 重复嵌套查询 #1010158081142619506

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -501,6 +505,155 @@ func (f *FormatFactory) SetData(data map[string]any) {
+			}
+		}
+	} else if len(termAggs) == 3 {
+		// 三维度场景（场景6）
```
你这个写法遇到 4 维度 5 维度的岂不是不能处理？

Zhaoyikaiii replied:
已修复

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -501,6 +505,155 @@ func (f *FormatFactory) SetData(data map[string]any) {
+	var nestedAggs []NestedAgg
+
+	for _, aggInfo := range f.aggInfoList {
+		switch aggInfo.(type) {
```
这个写法明显有问题，这样写的话，TimeAgg 的类型不就丢了吗？

Zhaoyikaiii replied:
已修复

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -501,6 +505,155 @@ func (f *FormatFactory) SetData(data map[string]any) {
+	ReverseAggName = "reverse_nested"
+)
+
+func (f *FormatFactory) reverseCheckAgg(aggregates metadata.Aggregates) []any {
```
你这个方法实现太复杂了，完全是 case by case 的适配，不是应该抽象成实现逻辑吗？

Zhaoyikaiii replied:
已修复

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -708,7 +869,7 @@ func (f *FormatFactory) EsAgg(aggregates metadata.Aggregates) (string, elastic.A
 		}
 	}
-
+	f.aggInfoList = f.reverseCheckAgg(aggregates)
```
这个方法建议放到 f.Agg() 函数里面，形成一个整体，因为如果不执行 reverseCheckAgg，f.Agg() 生产的 es dsl 一定会报错，所以他们俩本身就是一个整体，完全没必要拆分

Zhaoyikaiii replied:
已修复

Zhaoyikaiii committed:
fix: 重复嵌套查询 #1010158081142619506

Zhaoyikaiii committed:
feat: 增加嵌套查询的支持，优化查询逻辑和聚合处理

Adds support for nested queries in Elasticsearch, particularly for 'MustNot' conditions, by adjusting the query structure to ensure correct filtering.
Also, optimizes aggregation processing to handle nested fields correctly, ensuring accurate data aggregation in complex scenarios.

Zhaoyikaiii committed:
chore: 移除修复must_not必须在nested外面的代码

shamcleren committed:
Merge branch 'master' into fix/unify_query_nested_query/#1010158081142619506

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -501,14 +511,49 @@ func (f *FormatFactory) SetData(data map[string]any) {
-	for _, aggInfo := range f.aggInfoList {
+	newAggInfoList := f.nestedFieldCheckAgg()
+
+	for _, aggInfo := range newAggInfoList {
```
这里可以直接放到下面来，for _, aggInfo := range f.nestedFieldCheckAgg() {

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -501,14 +511,49 @@ func (f *FormatFactory) SetData(data map[string]any) {
 	mapData("", data, f.data)
 }
 
+func (f *FormatFactory) nestedFieldCheckAgg() (newAggInfoList aggInfoList) {
```
这个可以直接改名成为 aggInfoListWithNested，作为输出的数组直接调用

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -501,14 +511,49 @@ func (f *FormatFactory) SetData(data map[string]any) {
 }
 
+func (f *FormatFactory) nestedFieldCheckAgg() (newAggInfoList aggInfoList) {
+	nestedPathCheck := func(path string) bool {
```
这个方法，可以抽象到外层。作为 f.nestedPathCheck  来使用，增加复用度，以及注释说明这个方法的作用

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -501,14 +511,49 @@ func (f *FormatFactory) SetData(data map[string]any) {
+		return false
+	}
+
+	for i := len(f.aggInfoList) - 1; i >= 0; i-- {
```
这里增加注释，表明为什么要倒序

Zhaoyikaiii replied:
已添加

Zhaoyikaiii committed:
chore: 补充代码注释，修改函数命名

Zhaoyikaiii committed:
chore: 添加代码注释
